File: scraper_service/career_watcher.js
/**
 * career_watcher.js
 * 
 * Polls every company in watchlist.json. For each company it:
 *   1. Fetches the career page using a stealth headless browser
 *   2. Extracts all job listings (title + URL)
 *   3. Compares against a stored snapshot hash
 *   4. If NEW jobs are detected, outputs them as JSON to stdout
 * 
 * Usage: node career_watcher.js
 * The Python discovery layer calls this script and consumes stdout.
 */

const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
puppeteer.use(StealthPlugin());

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const WATCHLIST_PATH = path.join(__dirname, '..', 'watchlist.json');
const SNAPSHOTS_DIR = path.join(__dirname, 'snapshots');

// ─── Helpers ──────────────────────────────────────────────────────────────────

function getSnapshotPath(companyName) {
    const slug = companyName.toLowerCase().replace(/[^a-z0-9]/g, '_');
    return path.join(SNAPSHOTS_DIR, `${slug}.json`);
}

function loadSnapshot(companyName) {
    const p = getSnapshotPath(companyName);
    if (!fs.existsSync(p)) return { hash: null, jobs: [] };
    try {
        return JSON.parse(fs.readFileSync(p, 'utf8'));
    } catch {
        return { hash: null, jobs: [] };
    }
}

function saveSnapshot(companyName, hash, jobs) {
    const p = getSnapshotPath(companyName);
    fs.writeFileSync(p, JSON.stringify({ hash, jobs, updated_at: new Date().toISOString() }, null, 2));
}

function hashContent(str) {
    return crypto.createHash('md5').update(str).digest('hex');
}

// ─── Generic job list extractor ───────────────────────────────────────────────
// Tries multiple common selectors to find job links on any career page

async function extractJobs(page, companyUrl) {
    return await page.evaluate((baseUrl) => {
        const selectors = [
            'a[href*="job"]',
            'a[href*="position"]',
            'a[href*="career"]',
            'a[href*="opening"]',
            '[class*="job"] a',
            '[class*="position"] a',
            '[class*="role"] a',
            '[class*="listing"] a',
            '[data-job-id] a',
            'li a',
        ];

        const seen = new Set();
        const jobs = [];

        for (const sel of selectors) {
            const els = Array.from(document.querySelectorAll(sel));
            for (const el of els) {
                const title = el.innerText ? el.innerText.trim() : '';
                const href = el.href || '';
                if (title.length < 5 || title.length > 200) continue;
                if (!href || seen.has(href)) continue;
                // Filter out generic nav links
                const lower = title.toLowerCase();
                if (['home', 'about', 'contact', 'login', 'sign in', 'sign up', 'blog', 'help'].includes(lower)) continue;
                seen.add(href);
                jobs.push({ title, url: href });
            }
        }
        return jobs;
    }, companyUrl);
}

// ─── Main watcher loop ────────────────────────────────────────────────────────

async function watchAll() {
    if (!fs.existsSync(WATCHLIST_PATH)) {
        console.error('watchlist.json not found');
        process.exit(1);
    }

    const { companies } = JSON.parse(fs.readFileSync(WATCHLIST_PATH, 'utf8'));
    if (!fs.existsSync(SNAPSHOTS_DIR)) fs.mkdirSync(SNAPSHOTS_DIR, { recursive: true });

    const browser = await puppeteer.launch({
        headless: 'new',
        args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage']
    });

    const allNewJobs = [];

    for (const company of companies) {
        const { name, careers_url } = company;
        console.error(`[Watcher] Checking: ${name}`);
        const page = await browser.newPage();

        try {
            await page.setDefaultNavigationTimeout(30000);
            await page.goto(careers_url, { waitUntil: 'domcontentloaded' });
            // Extra wait for JS-heavy SPAs
            await new Promise(r => setTimeout(r, 3000));

            const jobs = await extractJobs(page, careers_url);

            if (jobs.length === 0) {
                console.error(`[Watcher] ${name}: No jobs extracted (may need custom selector)`);
                await page.close();
                continue;
            }

            // Build a stable hash from the sorted job titles+urls
            const contentStr = jobs.map(j => `${j.title}|${j.url}`).sort().join('\n');
            const liveHash = hashContent(contentStr);

            const snapshot = loadSnapshot(name);

            if (snapshot.hash === liveHash) {
                console.error(`[Watcher] ${name}: No changes (${jobs.length} jobs, hash unchanged)`);
            } else {
                // Find genuinely new jobs by comparing URLs
                const snapshotUrls = new Set((snapshot.jobs || []).map(j => j.url));
                const newJobs = snapshot.hash === null
                    ? [] // First run — just save baseline, don't flood the pipeline
                    : jobs.filter(j => !snapshotUrls.has(j.url));

                if (newJobs.length > 0) {
                    console.error(`[Watcher] ${name}: 🚨 ${newJobs.length} NEW job(s) detected!`);
                    for (const job of newJobs) {
                        allNewJobs.push({
                            title: job.title,
                            company: name,
                            url: job.url,
                            source: 'company_watcher',
                            location: 'India',
                        });
                    }
                } else {
                    console.error(`[Watcher] ${name}: Hash changed but no new unique URLs (page re-ordered or UI update)`);
                }

                saveSnapshot(name, liveHash, jobs);
            }
        } catch (err) {
            console.error(`[Watcher] ${name}: ERROR — ${err.message}`);
        } finally {
            await page.close();
        }
    }

    await browser.close();

    // Output new jobs as a single JSON line on stdout for Python to consume
    console.log(JSON.stringify(allNewJobs));
}

watchAll().catch(err => {
    console.error('Career watcher fatal error:', err);
    console.log('[]'); // Always output valid JSON so Python doesn't crash
    process.exit(1);
});

File: scraper_service/career_watcher.py
import json
import logging
import os
import sys
import time
import hashlib
from urllib.parse import urlparse
from playwright.sync_api import sync_playwright
from playwright_stealth import Stealth
from engine.utils import get_data_dir

logging.basicConfig(level=logging.ERROR, stream=sys.stderr)
logger = logging.getLogger(__name__)

def get_snapshot_path(company_name: str) -> str:
    import re
    slug = re.sub(r'[^a-z0-9]', '_', company_name.lower())
    snapshots_dir = os.path.join(get_data_dir(), 'snapshots')
    os.makedirs(snapshots_dir, exist_ok=True)
    return os.path.join(snapshots_dir, f"{slug}.json")

def load_snapshot(company_name: str) -> dict:
    p = get_snapshot_path(company_name)
    if not os.path.exists(p):
        return {"hash": None, "jobs": []}
    try:
        with open(p, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception:
        return {"hash": None, "jobs": []}

def save_snapshot(company_name: str, hash_val: str, jobs: list):
    p = get_snapshot_path(company_name)
    from datetime import datetime
    with open(p, 'w', encoding='utf-8') as f:
        json.dump({"hash": hash_val, "jobs": jobs, "updated_at": datetime.utcnow().isoformat()}, f, indent=2)

def hash_content(s: str) -> str:
    return hashlib.md5(s.encode('utf-8')).hexdigest()

def extract_jobs(page, company_url):
    return page.evaluate("""(baseUrl) => {
        const selectors = [
            'a[href*="job"]', 'a[href*="position"]', 'a[href*="career"]', 'a[href*="opening"]',
            '[class*="job"] a', '[class*="position"] a', '[class*="role"] a', '[class*="listing"] a',
            '[data-job-id] a', 'li a',
        ];
        const seen = new Set();
        const jobs = [];
        for (const sel of selectors) {
            const els = Array.from(document.querySelectorAll(sel));
            for (const el of els) {
                const title = el.innerText ? el.innerText.trim() : '';
                const href = el.href || '';
                if (title.length < 5 || title.length > 200) continue;
                if (!href || seen.has(href)) continue;
                const lower = title.toLowerCase();
                if (['home', 'about', 'contact', 'login', 'sign in', 'sign up', 'blog', 'help'].includes(lower)) continue;
                seen.add(href);
                jobs.push({ title, url: href });
            }
        }
        return jobs;
    }""", company_url)

def watch_all() -> list:
    watchlist_path = os.path.join(get_data_dir(), 'watchlist.json')
    if not os.path.exists(watchlist_path):
        logger.error('watchlist.json not found')
        sys.exit(1)
        
    try:
        with open(watchlist_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
            companies = data.get('companies', [])
    except Exception as e:
        logger.error(f'Error reading watchlist.json: {e}')
        sys.exit(1)
        
    all_new_jobs = []
    stealth = Stealth()
    
    try:
        with sync_playwright() as p:
            with p.chromium.launch(
                headless=True,
                args=['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage']
            ) as browser:
                for company in companies:
                    name = company.get('name')
                    careers_url = company.get('careers_url')
                    if not name or not careers_url:
                        continue
                        
                    logger.error(f"[Watcher] Checking: {name}")
                    context = browser.new_context()
                    page = context.new_page()
                    stealth.apply_stealth_sync(page)
                    
                    try:
                        page.goto(careers_url, wait_until='domcontentloaded', timeout=30000)
                        time.sleep(3)
                        
                        jobs = extract_jobs(page, careers_url)
                        
                        if not jobs:
                            logger.error(f"[Watcher] {name}: No jobs extracted (may need custom selector)")
                            continue
                            
                        # Sort jobs to create stable hash
                        sorted_jobs = sorted(jobs, key=lambda x: x['url'])
                        content_str = "\n".join(f"{j['title']}|{j['url']}" for j in sorted_jobs)
                        live_hash = hash_content(content_str)
                        
                        snapshot = load_snapshot(name)
                        
                        if snapshot.get('hash') == live_hash:
                            logger.error(f"[Watcher] {name}: No changes ({len(jobs)} jobs, hash unchanged)")
                        else:
                            snapshot_urls = {j['url'] for j in snapshot.get('jobs', [])}
                            if snapshot.get('hash') is None:
                                new_jobs = [] # First run — just save baseline
                            else:
                                new_jobs = [j for j in jobs if j['url'] not in snapshot_urls]
                                
                            if new_jobs:
                                logger.error(f"[Watcher] {name}: 🚨 {len(new_jobs)} NEW job(s) detected!")
                                for job in new_jobs:
                                    all_new_jobs.append({
                                        "title": job['title'],
                                        "company": name,
                                        "url": job['url'],
                                        "source": 'company_watcher',
                                        "location": 'India'
                                    })
                            else:
                                logger.error(f"[Watcher] {name}: Hash changed but no new unique URLs (page re-ordered or UI update)")
                                
                            save_snapshot(name, live_hash, jobs)
                            
                    except Exception as e:
                        logger.error(f"[Watcher] {name}: ERROR — {e}")
                    finally:
                        page.close()
    except Exception as e:
        logger.error(f"Career watcher fatal error: {e}")
        
    return all_new_jobs

if __name__ == "__main__":
    try:
        res = watch_all()
        print(json.dumps(res))
    except Exception as e:
        logger.error(f'Career watcher fatal error: {e}')
        print('[]')
        sys.exit(1)
